Add ROCm limitations to docs (#5107)
cupy/cupy: docs/source/install.rst

  If you don't specify the ``HCC_AMDGPU_TARGET`` environment variable, CuPy will be built for the GPU architectures available on the build host.
  This behavior is specific to ROCm builds; when building CuPy for NVIDIA CUDA, the build result is not affected by the host configuration.

Limitations
-----------

The following features are not available due to the limitation of ROCm or because that they are specific to CUDA:

* CUDA Array Interface
* cuTENSOR

> * Handling extremely large arrays whose size is around 32-bit boundary (HIP is known to fail with sizes 2**32-1024)

Author: ```suggestion
* Handling extremely large arrays whose size is around 32-bit boundary (HIP is known to fail with sizes `2**32-1024`)
```

* Sparse matrices (``cupyx.scipy.sparse``)
* cuDNN (hipDNN)
* Hermitian/symmetric eigenvalue solver (``cupy.linalg.eigh``)
* Polynomial roots (uses Hermitian/symmetric eigenvalue solver)

The following features may not work in edge cases (e.g., some combinations of dtype):

.. note::
   We are investigating the root causes of the issues. They are not necessarily
   CuPy's issues, but ROCm may have some potential bugs.

* ``cupy.ndarray.__getitem__`` (`#4653 <https://github.com/cupy/cupy/pull/4653>`_)
* ``cupy.ix_`` (`#4654 <https://github.com/cupy/cupy/pull/4654>`_)
* Some polynomial routines (`#4656 <https://github.com/cupy/cupy/pull/4656>`_, `#4658 <https://github.com/cupy/cupy/pull/4658>`_, `#4758 <https://github.com/cupy/cupy/pull/4758>`_, `#4759 <https://github.com/cupy/cupy/pull/4759>`_)
* ``cupy.broadcast`` (`#4662 <https://github.com/cupy/cupy/pull/4662>`_)
* ``cupy.convolve`` (`#4668 <https://github.com/cupy/cupy/pull/4668>`_)
* ``cupy.correlate`` (`#4781 <https://github.com/cupy/cupy/pull/4781>`_)
* Some random sampling routines (``cupy.random``, `#4770 <https://github.com/cupy/cupy/pull/4770>`_)
* ``cupy.linalg.einsum``

> * ``cupyx.scipy.ndimage`` and ``cupyx.scipy.signal``(`#4878 <https://github.com/cupy/cupy/pull/4878>`_, `#4879 <https://github.com/cupy/cupy/pull/4879>`_, `#4880 <https://github.com/cupy/cupy/pull/4880>`_)

Reviewer: Could it be this?
```suggestion
* ``cupyx.scipy.ndimage`` and ``cupyx.scipy.signal`` (`#4878 <https://github.com/cupy/cupy/pull/4878>`_, `#4879 <https://github.com/cupy/cupy/pull/4879>`_, `#4880 <https://github.com/cupy/cupy/pull/4880>`_)
```

Reviewer: https://techwriter.documatt.com/sphinx-errors/5-inline-literal-start-string-without-end-string.html#solution-to-2

Reviewer 2: Thanks!!

  If you don't specify the ``HCC_AMDGPU_TARGET`` environment variable, CuPy will be built for the GPU architectures available on the build host.
  This behavior is specific to ROCm builds; when building CuPy for NVIDIA CUDA, the build result is not affected by the host configuration.

Limitations
-----------

> The following CUDA-specific features are not available in ROCm:

Author: ```suggestion
The following features are not available due to the limitation of ROCm or they are specific to CUDA:
```

* CUDA Array Interface
* cuTENSOR

> * Atomic addition in FP16 (``cupy.ndarray.scatter_add`` and `cupyx.scatter_add`)

Author: ```suggestion
* Atomic addition in FP16 (``cupy.ndarray.scatter_add`` and ``cupyx.scatter_add``)
```

* Multi-GPU FFT and FFT callback
* Some random number generation algorithms

The following features are not yet supported:

* Several options in RawKernel/RawModule APIs: Jitify, dynamic parallelism
* Sparse matrices (``cupyx.scipy.sparse``)
* cuDNN (hipDNN)
* Hermitian/symmetric eigenvalue solver (``cupy.linalg.eigh``)
* Polynomial roots (uses Hermitian/symmetric eigenvalue solver)

The following features may not work in edge cases (e.g., some combinations of dtype):

.. note::
   We are investigating the root causes of the issues. They are not necessarily
   CuPy's issues, but ROCm may have some potential bugs.

> * 32-bit CArray boundary (HIP is known to fail with sizes 2**32-1024)

Author: Maybe move this line to the first section?
I also think it's better to reword this as `Handling extremely large arrays whose size is around 32-bit boundary (HIP is known to fail with sizes 2**32-1024)` as CArray is not familiar word for readers.

Reviewer 2: Will fix as such!

> * ``cupy.ndarray.__getitem__``

Author: Could you add some example failing cases? `__getitem__` sounds too broad.

Reviewer 2: Sure. I agree that ``__getitem__`` looks too broad. But the failing cases are actually as below and I also don't think  it helps the reader...

https://github.com/cupy/cupy/blob/a652e4707d7a16f431379bf8919c5a47cef4c990/tests/cupy_tests/core_tests/test_ndarray_indexing.py#L81-L94

Author: As discussed offline, let's add a link to the PR that added this skip code.

> * ``cupy.cuda.cufft.PlanNd``

Reviewer 3: I think this is currently private (despite I'd love to make it public)?

Reviewer 2: Thanks! I'll remove the line.

* ``cupy.ix_``
* Some polynomial routines
* ``cupy.broadcast``
* ``cupy.convolve``
* ``cupy.correlate``
* Some random sampling routines (``cupy.random``)

> * ``cupy.cublas.dgmm``

Author: I think only private APIs can be excluded. (is this called from some API?)

Reviewer 2: Let me check it shortly...

Reviewer 2: `dgmm` is not used from the public APIs so I'll remove the line!
